gene/mutate.go

```diff
@@ -167,59 +174,87 @@ func Insertion(g *Gene, genomePos int, alt []dna.Base) (EffectPrediction, error)
+		}
 
 		// Update cDNA
 		g.cdnaSeq = dna.Insert(g.cdnaSeq, int64(cdnaPos+1), alt)
 
-		//TODO: Effect Prediction WIP
-		//answer.AaPos = cdnaPos / 3
-		//
-		//if len(alt) % 3 != 0 { // Causes Frameshift // TODO this is a bit jury rigged, but it works....
-		//	answer.Consequence = Frameshift
-		//	currCodon, err = CdnaPosToCodon(g, cdnaPos + 1)
-		//	var stopFound bool = true
-		//	for currCdnaPos := cdnaPos + 1 + 3; dna.TranslateCodon(&currCodon) != dna.Stop; currCdnaPos += 3 {
-		//		answer.AaAlt = append(answer.AaAlt, dna.TranslateCodon(&currCodon))
-		//		if currCdnaPos + (3 - frame) > len(g.cdnaSeq) { // if do dont hit a stop by the end of the cdnaSeq, then move to genomic DNA
-		//			currGdnaPos := g.cdsEnds[len(g.cdsEnds)-1] + 1 - frame
-		//			for ;dna.TranslateSeq(g.genomeSeq[currGdnaPos:currGdnaPos + 3])[0] != dna.Stop; currGdnaPos += 3 {
-		//				answer.AaAlt = append(answer.AaAlt, dna.TranslateSeq(g.genomeSeq[currGdnaPos:currGdnaPos + 3])[0])
-		//				if currGdnaPos + 3 > len(g.genomeSeq) {
-		//					stopFound = false
-		//					err = errors.New("frameshift was unable to find a stop before the end of the gene")
-		//					break
-		//				}
-		//			}
-		//			break
-		//		} else {
-		//			currCodon, err = CdnaPosToCodon(g, currCdnaPos)
-		//		}
-		//	}
-		//	if stopFound {
-		//		answer.AaAlt = append(answer.AaAlt, dna.Stop)
-		//	}
-		//} else { // In-Frame
-		//	answer.Consequence = InFrameInsertion
-		//	if frame != 0 { // Disrupts an existing codon
-		//		answer.AaAlt = dna.TranslateSeq(g.cdnaSeq[(cdnaPos + 1) - frame : (cdnaPos + 1) + len(alt) + (3 - frame)])
-		//		if answer.AaRef[0] == answer.AaAlt[0] {
-		//			answer.AaRef = nil
-		//			answer.AaAlt = answer.AaAlt[1:]
-		//			answer.AaPos++
-		//		}
-		//	} else { // Does not disrupt existing codons AND is inframe
-		//		answer.AaAlt = dna.TranslateSeq(alt)
-		//	}
-		//}
+		answer.AaPos = cdnaPos / 3
+
+		if len(alt)%3 != 0 { // Causes Frameshift // TODO this is a bit jury rigged, but it works.
+			answer.StopDist = 0
+			answer.Consequence = Frameshift
+			currCodon, err = CdnaPosToCodon(g, cdnaPos+1)
+			//var stopFound bool = true
+			answer.AaAlt = dna.TranslateSeq(g.cdnaSeq[(cdnaPos+1)-frame : (cdnaPos+1)+len(alt)+(3-frame)])
+			var tmpCodon dna.Codon
+
+			for currCdnaPos := cdnaPos + 1 + 3; dna.TranslateCodon(&currCodon) != dna.Stop; currCdnaPos += 3 {
+				if answer.AaRef[0] == answer.AaAlt[0] {
+					answer.AaPos++
+					answer.AaRef[0] = g.protSeq[answer.AaPos]
+					answer.AaAlt = answer.AaAlt[1:]
+					if len(answer.AaAlt) == 0 {
+						tmpCodon, err = CdnaPosToCodon(g, currCdnaPos)
+						answer.AaAlt = append(answer.AaAlt, dna.TranslateCodon(&tmpCodon))
+					}
+				} else {
+					answer.StopDist++
+				}
+				//answer.AaAlt = append(answer.AaAlt, dna.TranslateCodon(&currCodon))
+				if currCdnaPos+(3-frame) > len(g.cdnaSeq) { // if you don't hit a stop by the end of the cdnaSeq, then move to genomic DNA
```

Comment:
I am confused about how you can go beyond the cDNA by looking at the genomic sequence.  I thought the genomic sequence stopped at the same place as the cDNA (in this struct).  I thought the genomeSeq was what was transcribed (transcription start to transcription end) and that the cdnaSeq was the transcript after splicing (no introns).

Reply:
I screwed up here. For starters I was moving to genomic DNA to find a stop codon which is completely incorrect as you point out. Another problem I realized is that what I am calling 'cDNA' is actually just the coding sequence (w/o UTRs). Proper cDNA includes the UTRs so I will need to rename this field. Instead of checking the genomic DNA I need to be checking the 3' UTR. This will take some work. Converting pull to a draft.